From 12ffaa8c34bf96265431697506d54636eeb916ed Mon Sep 17 00:00:00 2001
From: Codrin Ciubotariu <codrin.ciubotariu@microchip.com>
Date: Wed, 16 Mar 2022 16:16:30 +0200
Subject: [PATCH 057/471] ALSA: dmaengine_pcm: add workaround for MCHP-PDMC

Add workaround to clear PDMC's channel index when used as BE with ASRC.

Signed-off-by: Codrin Ciubotariu <codrin.ciubotariu@microchip.com>
[claudiu.beznea: adapt to v5.15.82]
Signed-off-by: Claudiu Beznea <claudiu.beznea@microchip.com>
---
 sound/core/pcm_dmaengine.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/sound/core/pcm_dmaengine.c b/sound/core/pcm_dmaengine.c
index 0fe93b423c4e..6998af4a1eb9 100644
--- a/sound/core/pcm_dmaengine.c
+++ b/sound/core/pcm_dmaengine.c
@@ -136,6 +136,21 @@ static void dmaengine_pcm_dma_complete(void *arg)
 	struct snd_pcm_substream *substream = arg;
 	struct dmaengine_pcm_runtime_data *prtd = substream_to_prtd(substream);
 
+	/*
+	 * workaround to clear mchp-pdmc's channel index when used as BE with
+	 * mchp-asrc
+	 */
+	if (substream->pcm->internal &&
+	    !strncmp(prtd->dma_chan->slave->driver->name, "mchp-pdmc", sizeof("mchp-pdmc"))) {
+		unsigned int sample_size = samples_to_bytes(substream->runtime, 1);
+		u8 *dma_ptr = substream->runtime->dma_area + prtd->pos;
+		u8 *dma_ptr_end = dma_ptr + snd_pcm_lib_period_bytes(substream);
+
+		for (; dma_ptr < dma_ptr_end; dma_ptr += sample_size)
+			*dma_ptr = 0;
+	}
+
+	
 	new_pos = prtd->pos + snd_pcm_lib_period_bytes(substream);
 	if (new_pos >= snd_pcm_lib_buffer_bytes(substream))
 		new_pos = 0;
-- 
2.34.1

